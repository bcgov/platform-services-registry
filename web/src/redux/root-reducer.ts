import { combineReducers } from 'redux';
import githubIDReducer from './githubID/githubID.reducer';

export default combineReducers({
  githubID: githubIDReducer,
});
