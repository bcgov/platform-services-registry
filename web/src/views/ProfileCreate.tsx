//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { useKeycloak } from '@react-keycloak/web';
import React, { useEffect, useState } from 'react';
import { Redirect } from 'react-router-dom';
import CreateFormMetadata from '../components/profileCreate/CreateFormMetadata';
import CreateFormPO from '../components/profileCreate/CreateFormPO';
import CreateFormProject from '../components/profileCreate/CreateFormProject';
import CreateFormRequest from '../components/profileCreate/CreateFormRequest';
import CreateFormTC from '../components/profileCreate/CreateFormTC';
import { ROUTE_PATHS } from '../constants';
import useCommonState from '../hooks/useCommonState';
import useRegistryApi from '../hooks/useRegistryApi';
import { promptErrToastWithText, promptSuccessToastWithText } from '../utils/promptToastHelper';
import { transformForm } from '../utils/transformDataHelper';
import Wizard, { WizardPage } from '../utils/Wizard';

const ProfileCreate: React.FC = () => {
  const api = useRegistryApi();
  const { keycloak } = useKeycloak();
  const { setOpenBackdrop } = useCommonState();

  const [ministry, setMinistry] = useState<any>([]);
  const [cluster, setCluster] = useState<any>([]);
  const [goBackToDashboard, setGoBackToDashboard] = useState(false);

  const onSubmit = async (formData: any) => {
    const { profile, productOwner, technicalContact, clusters } = transformForm(formData);
    setOpenBackdrop(true);
    try {
      // 1. Create the project profile.
      const response: any = await api.createProfile(profile);
      const profileId = response.data.id;

      // 2. Create contacts.
      const po: any = await api.createContact(productOwner);
      const tc: any = await api.createContact(technicalContact);

      // 3. Link the contacts to the profile.
      await api.linkContactToProfileById(profileId, po.data.id);
      await api.linkContactToProfileById(profileId, tc.data.id);

      // 4. Trigger provisioning
      await api.createNamespaceByProfileId(profileId, clusters);

      // 4. Create Project Request
      await api.createProjectRequestByProfileId(profileId);

      setOpenBackdrop(false);
      setGoBackToDashboard(true);
      // 5.All good? Tell the user.
      promptSuccessToastWithText('Your namespace request was successful');
    } catch (err) {
      setOpenBackdrop(false);
      const msg = `Unable to submit request at this time, reason = ${err.message}`;
      promptErrToastWithText(msg);
      console.log(err);
    }
  };

  useEffect(() => {
    async function wrap() {
      const ministryResponse = await api.getMinistry();
      const clusterResponse = await api.getCluster();
      setMinistry(ministryResponse.data);
      setCluster(clusterResponse.data);
    }
    wrap();
    // eslint-disable-next-line
  }, [keycloak]);

  if (goBackToDashboard) {
    return <Redirect to={ROUTE_PATHS.DASHBOARD} />;
  }
  return (
    <Wizard onSubmit={onSubmit}>
      <WizardPage>
        <CreateFormProject ministry={ministry} cluster={cluster} />
      </WizardPage>
      <WizardPage>
        <CreateFormMetadata />
      </WizardPage>
      <WizardPage>
        <CreateFormPO />
      </WizardPage>
      <WizardPage>
        <CreateFormTC />
      </WizardPage>
      <WizardPage>
        <CreateFormRequest />
      </WizardPage>
    </Wizard>
  );
};

export default ProfileCreate;
