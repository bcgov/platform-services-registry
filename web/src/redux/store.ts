import { createStore, applyMiddleware } from 'redux';
import logger from 'redux-logger';
import thunk from 'redux-thunk';
import { persistStore } from 'redux-persist';
import rootReducer from './root-reducer';

// TODO: thunk or saga?

const middlewares = [];
middlewares.push(thunk);
if (process.env.NODE_ENV === 'development') {
  middlewares.push(logger);
}

export const store = createStore(rootReducer, applyMiddleware(...middlewares));
export const persistor = persistStore(store);
