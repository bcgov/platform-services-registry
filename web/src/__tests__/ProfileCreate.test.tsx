//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { act } from "react-dom/test-utils";
// import { create } from "react-test-renderer";
import React from 'react';
import { Provider } from 'react-redux';
import { store } from '../redux/store';
import ProfileCreate from '../views/ProfileCreate';
import { MsalReactTester } from 'msal-react-tester';
import { MsalProvider } from '@azure/msal-react';

jest.mock(
  '../utils/getDecodedToken',
  () =>
    function getDecodedToken() {
      return {
        email: 'test@example.com',
        family_name: 'Jane',
        given_name: 'Doe',
        name: 'Jane Doe',
        preferred_username: 'janedoe@idir',
      };
    },
);

test('matches the snapshot', async() => {
  const stubPropCB = jest.fn();
  const msalTester = new MsalReactTester(); 
  msalTester.spyMsal();
  let container;
  msalTester.isNotLogged();

      // container = create(
      //   <MsalProvider instance={msalTester.client}>
      //     <Provider store={store}>
      //       <ProfileCreate openBackdropCB={stubPropCB} closeBackdropCB={stubPropCB} container />,
      //     </Provider>,
      //   </MsalProvider>
      // );

 // expect(container).toMatchSnapshot();
  msalTester.resetSpyMsal();
});
