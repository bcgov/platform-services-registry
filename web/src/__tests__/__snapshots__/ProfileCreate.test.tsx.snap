// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <form>
    <div
      class="css-hdrfx9"
    >
      <div
        class="css-1pslfv9"
      >
        <div>
          <div>
            <h1
              class="css-zi8sy1"
            >
              Tell us about your project
            </h1>
            <h3
              class="css-ao0kz4"
            >
              If this is your first time on the OpenShift platform you need to book an alignment meeting with the Platform Services team; Reach out to
               
              <a
                href="mailto:olena.mitovska@gov.bc.ca"
                rel="noopener noreferrer"
                target="_blank"
              >
                olena.mitovska@gov.bc.ca
              </a>
               
              to get started.
            </h3>
            <div
              class="css-l3rx45"
            >
              <label
                class="css-knfhfe"
                for="project-name"
              >
                Name
              </label>
              <div
                class=""
              >
                <input
                  class="css-4y926w"
                  name="project-name"
                  placeholder="Project X"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="css-l3rx45"
            >
              <label
                class="css-knfhfe"
                for="project-description"
              >
                Description
              </label>
              <div
                class=""
              >
                <textarea
                  class="css-4y926w"
                  name="project-description"
                  placeholder="A cutting edge web platform that enables Citizens to ..."
                  rows="5"
                />
              </div>
            </div>
            <div
              class="css-18m6ozg"
            >
              <label
                class="css-toc01s"
              >
                Is this a Priority Application?
              </label>
              <div
                class="css-xtali0"
              >
                <label
                  class="css-iw3ckw"
                >
                  <div
                    class="css-vurnku"
                  >
                    <input
                      class="css-p19g2b"
                      name="project-prioritySystem"
                      type="checkbox"
                    />
                    <svg
                      aria-hidden="true"
                      class="css-1orewpb"
                      fill="currentcolor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
                      />
                    </svg>
                    <svg
                      aria-hidden="true"
                      class="css-6gju6b"
                      fill="currentcolor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                      />
                    </svg>
                  </div>
                </label>
              </div>
            </div>
            <div
              class="css-18m6ozg"
            >
              <label
                class="css-toc01s"
              >
                Ministry Sponsor
              </label>
              <div
                class="css-xtali0"
                name="project-busOrgId"
              >
                <div
                  class=""
                  style="flex: 1;"
                >
                  <div
                    class="css-4cffwv"
                  >
                    <select
                      class="css-4y926w"
                      name="project-busOrgId"
                    >
                      <option>
                         Select... 
                      </option>
                    </select>
                    <svg
                      class="css-6ibwe0"
                      fill="currentcolor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="css-18m6ozg"
            >
              <label
                class="css-toc01s"
              >
                Cluster Name
              </label>
              <div
                class="css-xtali0"
                name="project-primaryClusterName"
              >
                <div
                  class=""
                  style="flex: 1;"
                >
                  <div
                    class="css-4cffwv"
                  >
                    <select
                      class="css-4y926w"
                      name="project-primaryClusterName"
                    >
                      <option>
                         Select... 
                      </option>
                    </select>
                    <svg
                      class="css-6ibwe0"
                      fill="currentcolor"
                      viewBox="0 0 24 24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="buttons"
        >
          <button
            class="misc-class-m-form-submit-btn css-1cr4ss0"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
